src/main/java/org/opensearch/neuralsearch/sparse/NeuralSparseIndexShard.java

```
@@ -0,0 +1,268 @@
+            return new CacheGatedForwardIndexReader(cacheItem.getReader(), cacheItem.getWriter((ramBytesUsed) -> {
+                throw new CircuitBreakingException("Circuit Breaker reaches limit", CircuitBreaker.Durability.PERMANENT);
+            }), sparseBinaryDocValues);
+        } else {
+            // For regular BinaryDocValues, create a SparseVectorReader wrapper
+            final SparseVectorReader luceneReader = docId -> {
+                try {
+                    if (binaryDocValues.advanceExact(docId)) {
+                        return new SparseVector(binaryDocValues.binaryValue());
+                    }
+                } catch (IOException e) {
+                    log.error("Failed to read doc {}", docId, e);
+                }
+                return null;
+            };
+
+            ForwardIndexCacheItem cacheItem = ForwardIndexCache.getInstance().getOrCreate(key, docCount);
+            return new CacheGatedForwardIndexReader(null, cacheItem.getWriter((ramBytesUsed) -> {
+                throw new CircuitBreakingException("Circuit Breaker reaches limit", CircuitBreaker.Durability.PERMANENT);
+            }), luceneReader);
+        }
+    }
+
+    private Set<FieldInfo> collectSparseFieldInfos(LeafReader leafReader) {
+        return StreamSupport.stream(leafReader.getFieldInfos().spliterator(), false)
+            .filter(SparseTokensField::isSparseField)
+            .collect(Collectors.toSet());
+    }
+
+    private void warmUpForwardIndex(BinaryDocValues binaryDocValues, CacheGatedForwardIndexReader cacheGatedForwardIndexReader)
+        throws IOException {
+        int docId = binaryDocValues.nextDoc();
+        while (docId != DocIdSetIterator.NO_MORE_DOCS) {
+            try {
+                cacheGatedForwardIndexReader.read(docId);
+            } catch (IOException e) {
+                log.warn("[Neural Sparse] Failed to read doc {} during warm up", docId, e);
```

you swallow the exception here, that means if you run into IOException and fail to read one doc, but you try to read following docs

---

Should we stop? I think we can try to read as many docs as we can, just showing this information when there is a problem?

---

I checked Effective Java, it says that we should expect recovery with Checked Exception. So we swallow the exception here and resume next doc